connaisseur/sigstore_validator.py

```diff
@@ -0,0 +1,76 @@
+import json
+import logging
+import re
+import subprocess  # nosec
+
+from connaisseur.exceptions import (
+    CosignError,
+    CosignTimeout,
+    NotFoundException,
+    ValidationError,
+)
+
+
+def get_cosign_validated_digests(image: str, pubkey: str):
+    """
+    Gets and processes cosign validation output for a given `image` and `pubkey`
+    and either returns a list of valid digests or raises a suitable exception
+    in case no valid signature is found or cosign fails.
+    """
+    failed, stdout, stderr = invoke_cosign(image, pubkey)
+    logging.info(
+        "COSIGN output for image: %s; RETURNCODE: %s; STDOUT: %s; STDERR: %s",
+        image,
+        failed,
+        stdout,
+        stderr,
+    )
+    digests = []
+    if not failed:
+        for sig in stdout.splitlines():
+            try:
+                sig_data = json.loads(sig)
+                digests.append(sig_data["Critical"]["Image"]["Docker-manifest-digest"])
+            except json.JSONDecodeError:
+                logging.info("Non-json signature data from Cosign: %s", sig)
+                pass
+    elif stderr == "error: no matching signatures:\nunable to verify signature\n":
```

Comment:
Do these errors differ in their return codes? Maybe we could go for the return code as it might be stabler than the message itself.

Reply:
I wanted to keep it compatible with the standard connaisseur error messages to make it more understandable once we have multiple validators

Reply:
That's not what I meant, sorry. I was referring to line 37 and the error message that comes from `cosign`. If the binary returns different error codes we could use these instead of the error message to avoid having to change code if they change their error messages (which I could imagine to change rather than the codes)

Reply:
Gotcha. There is unfortunately no separate error codes. The output is still a bit hard to parse...

Reply:
alrighty then :-)